presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationBelowOuterJoin.java

```diff
@@ -134,8 +153,56 @@ private static boolean groupsOnAllOuterTableColumns(AggregationNode node, JoinNo
 
         private static PlanNode getFirstNonIdentityNode(PlanNode node)
         {
-            PlanNodeSearcher searcher = new PlanNodeSearcher(node).where(planNode -> !(planNode instanceof ProjectNode) || !((ProjectNode) planNode).isIdentity());
-            return searcher.findFirst().orElseThrow(() -> new IllegalStateException("no nodes below identity project"));
+            return searchFrom(node)
+                    .where(planNode -> !(planNode instanceof ProjectNode) || !((ProjectNode) planNode).isIdentity())
+                    .findFirst()
+                    .orElseThrow(() -> new IllegalStateException("no nodes below identity project"));
+        }
+
+        private static Optional<JoinNode> getFirstJoinNode(PlanNode node)
+        {
+            return searchFrom(node).skipOnlyWhen(planNode -> (planNode instanceof ProjectNode) && ((ProjectNode) planNode).isIdentity())
+                    .where(isInstanceOfAny(JoinNode.class))
+                    .findFirst();
+        }
+
+        private PlanNode addCoalesceIfNecessary(AggregationNode node, PlanNode result)
+        {
+            // The count functions return zero when the only input is null
+            List<Symbol> coalesceToZero = new ArrayList<>();
+            List<Symbol> coalesceToOne = new ArrayList<>();
+            List<String> zeroOnNullAggregations = ImmutableList.of("count_if", "approx_distinct");
```

Comment:
I think you need to maintain the separated list for each case. When aggregate returns `null`, `0`, `1` and any other value for empty set. 
It is due that user may add custom aggregate function. Maybe this information could be managed by FunctionRegistry?